lib/dma.h

```diff
@@ -164,15 +164,44 @@ bit_to_u8off(size_t val)
     return val % (CHAR_BIT);
 }
 
+static inline void
+u8_set_all(uint8_t *bm)
+{
+    *bm = ~0;
+}
+
+static inline void
+mask_any_lower_bits(uint8_t *byte, size_t start_idx)
+{
+    if (bit_to_u8off(start_idx) != 0) {
+        *byte &= ~((1 << bit_to_u8off(start_idx)) - 1);
+    }
+}
+
+static inline void
+mask_any_higher_bits(uint8_t *byte, size_t end_idx)
+{
+    if (bit_to_u8off(end_idx) != 0) {
+        *byte &= ((1 << bit_to_u8off(end_idx)) - 1);
+    }
+}
+
+static inline void
+apply_byte_mask(uint8_t *bitmap, uint8_t byte_mask)
+{
+    __atomic_or_fetch(bitmap, byte_mask, __ATOMIC_SEQ_CST);
+}
+
 static inline void
 _dma_mark_dirty(const dma_controller_t *dma, const dma_memory_region_t *region,
                 dma_sg_t *sg)
 {
-    size_t index;
-    size_t end;
+    size_t bm_sg_start;
```

Comment:
I'm not quite following what the "sg" part of this is - we already converted from sg in the range_to_pages() call.

Reply:
Agreed - not a great name. Maybe we call it bm_start_byte? I think it's worth emphasizing that this is the first byte of the sg list.

Reply:
works for me